doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -144,61 +144,129 @@ presentation of Cic is given in Chapter :ref:`calculusofinductiveconstructions`.
 
+.. insertgram term fixannot
+
 .. productionlist:: coq
-   term             : forall `binders` , `term`
-                    : fun `binders` => `term`
-                    : fix `fix_bodies`
-                    : cofix `cofix_bodies`
-                    : let `ident` [`binders`] [: `term`] := `term` in `term`
-                    : let fix `fix_body` in `term`
-                    : let cofix `cofix_body` in `term`
-                    : let ( [`name` , … , `name`] ) [`dep_ret_type`] := `term` in `term`
-                    : let ' `pattern` [in `term`] := `term` [`return_type`] in `term`
-                    : if `term` [`dep_ret_type`] then `term` else `term`
-                    : `term` : `term`
-                    : `term` <: `term`
-                    : `term` :>
-                    : `term` -> `term`
-                    : `term` `arg` … `arg`
-                    : @ `qualid` [`term` … `term`]
-                    : `term` % `ident`
-                    : match `match_item` , … , `match_item` [`return_type`] with
-                    :   [[|] `equation` | … | `equation`] end
-                    : `qualid`
-                    : `sort`
-                    : `num`
-                    : _
-                    : ( `term` )
-   arg              : `term`
-                    : ( `ident` := `term` )
-   binders          : `binder` … `binder`
-   binder           : `name`
-                    : ( `name` … `name` : `term` )
-                    : ( `name` [: `term`] := `term` )
-                    : ' `pattern`
-   name             : `ident` | _
-   qualid           : `ident` | `qualid` `field`
-   sort             : SProp | Prop | Set | Type
-   fix_bodies       : `fix_body`
-                    : `fix_body` with `fix_body` with … with `fix_body` for `ident`
-   cofix_bodies     : `cofix_body`
-                    : `cofix_body` with `cofix_body` with … with `cofix_body` for `ident`
-   fix_body         : `ident` `binders` [`annotation`] [: `term`] := `term`
-   cofix_body       : `ident` [`binders`] [: `term`] := `term`
-   annotation       : { struct `ident` }
-   match_item       : `term` [as `name`] [in `qualid` [`pattern` … `pattern`]]
-   dep_ret_type     : [as `name`] `return_type`
-   return_type      : return `term`
-   equation         : `mult_pattern` | … | `mult_pattern` => `term`
-   mult_pattern     : `pattern` , … , `pattern`
-   pattern          : `qualid` `pattern` … `pattern`
-                    : @ `qualid` `pattern` … `pattern`
-                    : `pattern` as `ident`
-                    : `pattern` % `ident`
-                    : `qualid`
-                    : _
-                    : `num`
-                    : ( `pattern` | … | `pattern` )
+   term : `binder_constr`
+        : `term100`
+   term100 : `term_cast`
+           : `term10` :>
+           : `term10`
+   term10 : `term9` `appl_arg_list`
+          : @ `qualid` `instance` `term9_list_opt`
+          : @ ? `ident` `ident_list`
```

Comment:
While technically part of the term syntax, this construction with `?ident` is reserved to "patterns".

```
Check(@?x n). (* Metavariables allowed only in patterns. *)
```

Would it be possible to split this (and other similar rules) as "extension rules" (similar to what plugins define)?

BTW, the space between `?` and `ident` is incorrect (won't parse if it exists).

Reply:
It's customary--and I think people expect--that all the productions for a nonterminal are together.  We can add little comments in the grammar for semantic restrictions, e.g. I can add " (* only for patterns *)" to the right of the production.

While it's possible to split this, you can't reuse the same nonterminal name twice in the document, even in different productionlists in different .rst files.  So one of them would have to be a prodn.

> BTW, the space between ? and ident is incorrect (won't parse if it exists).

A space in a production separates adjacent symbols.  It doesn't mean you have to put a space between the two symbols.  The grammar alone can't convey all these details.  You can't have spaces in a `num`, either:

```
digit : 0 .. 9
num   : num digit
      | digit
```

Reply:
>  So one of them would have to be a prodn.

Yes, that was my suggestion. You should view this pattern grammar as an extension on top of the base Gallina grammar (which is documented in the Gallina extensions chapter, for once it is in the right place).



> A space in a production separates adjacent symbols. It doesn't mean you have to put a space between the two symbols.

Right, except most of the time, you are *allowed* to insert a space, but here you are not. I guess that's something else that will be easier to fix when we switch to an all `prodn`-style grammar.

Reply:
I added a special case in Ocaml.